### ISSUE TITLE
preferred_side seems to have no effect on /sources_to_targets matrix API

### ISSUE THREAD

**Issue by klausfrandsen** [valhalla/valhalla#4440]
Hi there, 

First of all, thanks for an amazing piece of software! Without it, Our small startup would be in a completely different situation. I've encountered an issue though, that I hope you can help me resolve.

I have created this minimal example where I would expect the resulting matrix to be symmetric (or close to).

```
{
    "locations": [
    {
        "lat":56.611688,
        "lon":8.732719,
        "preferred_side": "either"
    },
    {
        "lat":56.611688,
        "lon":8.732719,
        "preferred_side": "either"
    }, 
    "costing": "auto",
}
```
The resulting matrices are:
duration: `[[0, 7], [268, 0]]`
distance:  `[[0.0, 124.0], [2860.0, 0.0]]`

The two points are on the same road 124 meters apart, and rather close to the road centerline. Here's an image from Google earth:
![image](https://github.com/valhalla/valhalla/assets/12632121/126b38da-1afc-4264-bab2-a2ea9df06ff6)

I have read this issue [1] about the administrative database. But as I'm trying to get it to ignore 'side', I would expect a missing administrative DB to not be the issue?

I have also read this post [2] about oneway streets. Which obviously also makes sense. And the street in question (Ålbækvej in DK) is not oneway.

[1]:
https://github.com/valhalla/valhalla/issues/4028

[2]:
https://github.com/valhalla/valhalla/issues/2686
 

Thanks a lot in advance for any help or pointers you might be able to provide,
Klaus

**Comment by klausfrandsen** [comment]
Thanks, I hadn't, but now I have. The info I can find seems rather straight forward as shown below

![image](https://github.com/valhalla/valhalla/assets/12632121/34c57dfe-d914-4764-b7f4-5d5d9038e327)

Anything speciel I should be looking for? Or are there other OSM tags than the ones shown here somewhere else that might can influence this?

**Comment by klausfrandsen** [comment]
I just tried using the /route API instead with the same two locations, and here the duration and distance is 7 sec and 124 meters, both when going from loc1 to loc2 and vice-versa. And this is true both with and without preferred_side set to "either".

Hope it helps to pin-point what I might be doing wrong in the matrix API.

Thanks

**Comment by nilsnolde** [comment]
When reporting issues, please always build and try latest master first. Also yes, do build the admin database (ideally from the planet), but it has nothing to do with this.

**Comment by klausfrandsen** [comment]
Sure thing. Sorry about that. I will report back here with the results.

Thanks for you help so far.

**Comment by klausfrandsen** [comment]
I have now build with latest master of gis-ops/docker-valhalla using latest master of valhalla/valhalla (git sha 98f06be from yesterday). The issue persists. 

I have also included the admins DB (build for Denmark only). That did not have an effect either.

However, I did some more experiments with the exact coordinates: 
The coordinates are from DAWA (Danish public addresses government service), but they are what they refer to as "vejpunkter", so coordinates on the centerline of the road, where you can access the address in question. However, if I change to using the address-coordinates which often pin-point the center of the building for the address (more or less), then Valhalla returns a much more sensible distance/duration matrices. Then, I tried to gradually move the waypoints closer from the address-coordinates towards the road-coordinates, and when I get below 5 meters from the road-coordinates, the non-sensible results appear again.

So in short, this request using coordinates lying 5 meters of the side of the road-centerline, returns the expected result where the distance matrix is close to symmetrical:

```
{
    "locations": [
    {
        "lat":56.6116874,
        "lon":8.7328333,
    },
    {
        "lat":56.6127584,
        "lon":8.7328487,
    }, 
    "costing": "auto",
}
```
duration: `[[0, 14], [238, 0]]`
distance: `[[0.0, 129.0], [157.0, 0.0]]`

**Comment by nilsnolde** [comment]
Huh, that definitely sounds like a side of street issue! I did a whole lot of work on matrix recently and kinda expected this one to go away with you updating. If I have a minute in the next days I’ll see what causes that!

**Comment by klausfrandsen** [comment]
Hi again, 

Just want to mention that I have identified a similar issue on another of Our routes. So if this is really a bug in Valhalla, it looks to be somewhat prevalent. 

In this case the two locations lie next to each other at the end of a residential road. 

Using road centerline coordinates:
```
{'lat': 55.743853, 'lon': 9.40724}
{'lat': 55.743861, 'lon': 9.40731}
Duration matrix: [[0, 201], [201, 0]]
Distance matrix: [[0.0, 1102.0], [1102.0, 0.0]]
```

Using coordinates pulled 7 meters towards the address:
```
{'lat': 55.7439096, 'lon': 9.4072899}
{'lat': 55.7438012, 'lon': 9.4073468}
Duration matrix: [[0, 0], [0, 0]]
Distance matrix: [[0.0, 0.0], [0.0, 0.0]]
```

**Comment by kevinkreiser** [comment]
some things to check

- are the routes trivial, do they take place on a single edge? this could explain wildly different expansions depending on which matrix algorithm you use. not all algorithms properly handle trivial (single edge routes) and because its matrix we have no easy way to fall back to one that does (since most pairs wont be trivial or at least we assume that)
- is node snapping coming into play, are the input locations close enough to an intersection that we use the intersection as the "vejpunkt"? (when this happens we use all the edges at the intersection)

using `either` for side of street is the default and even without an admin db it would be fine as it just means we wont filter either direction of the road from usage.

in your earlier example when you get it too close to the road center line i wonder if its node snapping. i would try disabling node snapping by setting node_snap_tolerance to 0 and see if that one is fixed.

in your last example the closest point on the road is probably the same for both addresses and so it becomes a zero length route, hence the 0s.

**Comment by klausfrandsen** [comment]
Thanks a lot for the input. Our use cases for the matrix lookups is a mix of longer routes and very short/trivial ones, so that sounds very plausible. Below are my findings based on your input. 

setting `node_snap_tolerance` to 0 indeed did solve my first example. Now, when using center-line coordinates I get a perfectly symmetric result:
```
{'lat': 56.6116874, 'lon': 8.7328333, 'node_snap_tolerance': 0}
{'lat': 56.6127584, 'lon': 8.7328487, 'node_snap_tolerance': 0}
Duration matrix: [[0, 8], [8, 0]]
Distance matrix: [[0.0, 123.0], [123.0, 0.0]]
```
So, just for my understanding, why does the matrix API not work well with waypoints that start at a node? You write "_(when this happens we use all the edges at the intersection)_", but what does that actually mean?

For the second example, setting `node_snap_tolerance` to 0 does not solve it. I assume this is because both the locations (Frihedslunden 11 and 13, 7300 Jelling, Denmark) lie at the end of an edge, and the end-node is therefore closest whether we snap or not. But I have tried to change the `source_to_target_algorithm` to `costmatrix`, and that (together with `node_snap_tolerance = 0` gives the following sensible result:

```
{'lat': 55.743853, 'lon': 9.40724, 'node_snap_tolerance': 0}
{'lat': 55.743861, 'lon': 9.40731, 'node_snap_tolerance': 0}
Duration matrix: [[0, 0], [0, 0]]
Distance matrix: [[0.0, 4.0], [4.0, 0.0]]
```
So can use `source_to_target_algorithm = costmatrix` in general, or will that have other consequences for e.g. longer routes?


Thanks

**Comment by kevinkreiser** [comment]
> For the second example, setting node_snap_tolerance to 0 does not solve it.

yeah i didnt expect it to i was merely explaining that what happens is we have to use the point on the road geometry closest to your input point, when that point is the same point for two inputs then of course the solution to the route will be a 0 length route.

in my opinion you should always use the "vejpunkter" as they should be the locations to which you want to actually route. i think there is still a bug here with trivial routes and using all the edges at a node though. by all the edges at a node i mean when we node snap we move the location to the intersection in the graph and then we consider using all the edges that are attached to that intersection. this is helpful depending on your use-case. as you can imagine we have to kind of guess which use case someone has and try to pick a reasonable default but it seems in this case the algorithm itself is tripped up by this input. we'll need to debug it.

the best think we could do is take your failing example above and model it in testing tools (gurka) to try to reproduce it. then we can step through the code and see whats going wrong.

using costmatrix will have better performance in terms of latency than TDMatrix but it will have worse memory consumption. also costmatrix should do worse with trivial routes..

i guess im saying im sorry we need to look into this a bit more we're just scratching the surface of the problem.

**Comment by klausfrandsen** [comment]
Thanks, I'll appreciate it if you have time to debug it.


> when that point is the same point for two inputs then of course the solution to the route will be a 0 length route.

Be aware that the returned distance matrix in my previous post is not [0 0][0 0]. It is returning 4 meters. (just in case you mis-read that detail, at it sounds like you assume a 0 length route)

**Comment by nilsnolde** [comment]
I should've mentioned that I only worked on CostMatrix recently. So that'd make sense. There definitely is an issue with trivial routes there too, but that shouldn't matter really unless you get the shape (and even there I tried to band-aid). Might be that TDMatrix has a bug there. It's fine to use CostMatrix all the way for many:many requests, more memory, but vastly faster execution.

**Comment by klausfrandsen** [comment]
So, I have digged some more into this, and to me it seems like a rather prevalent issue when routing on shorter distances. Below, I have a snippet from a larger matrix, showing 20 waypoints in a neighbourhood. To be able to identify the issues "automatically" I have written a small routine that checks if any edge (e.g. A->B) in the cost matrix is longer than any of the possible two-edges that goes via another waypoint (e.g. A->C + C->B). In the ideal world there should be no such "shortcuts" via another waypoint, and to me they must indicate an error of some sort, as going from A->B should never be longer/slower than going from A->C + C->B. 

In the example below, it identifies 30 of those cases (out of 20x20=400 costs) for the duration-matrix, where it is 20% or more faster to take a "shortcut" via another node. I have drawn two of the worst examples onto the map below.

<img width="600" alt="image" src="https://github.com/valhalla/valhalla/assets/12632121/4f374a5a-32fc-4fc7-b184-2e9d5b9c35cd">

(just ignore the blue lines, as they are not relevant in this context)

Also note that for these two examples drawn, the waypoints involved are not close to any road-nodes, so node-snapping should not be the issue here I assume. `node_snap_tolerance` is still set to 0 as recommended earlier in this thread.

This is for `source_to_target_algorithm = costmatrix`. For `source_to_target_algorithm = timedistancematrix` the overall result is the same with only minor differences in the costs returned.

Let me know if you want the complete example for debugging in one form or another, or if there are any settings I can try or logs I can enable and send to you, to aid in debugging.

I hope this helps. Thanks for your support
Klaus

**Comment by nilsnolde** [comment]
Thanks for the details @klausfrandsen ! What you're seeing is definitely possible while being valid. Admittedly the examples you give are quite extreme. It's possible though that you can arrive at a location on an edge which you wouldn't be able to depart from, hence leading to such "inconsistencies", e.g. one-ways. That shouldn't happen if there are node snaps, but it can happen for edge snaps (which you're kinda forcing). 

I'd like to look at this a bit more though. Can you paste the request/locations for the 2 examples you're drawing there? That'd be immensely helpful. Thanks!

**Comment by klausfrandsen** [comment]
Hi Nils

Sorry for the late reply. 

Below are the coordinates and duration cost matrices covering the two examples with red arrows in the previous image. (Note that there was a small bug in the upper red arrows. I have updated the image in the above post to fix this.)

```
#### Matrix duration_matrix_106_37_33 ####
     [0]  [1]  [2] 
[0]     0   14    7   |  (55.515759, 9.527645) - Marielystvej 106, 6000 Kolding, Denmark
[1]   207    0   10   |  (55.516036, 9.525968) - Marielystvej 37, 6000 Kolding, Denmark
[2]     7    5    0   |  (55.515759, 9.526499) - Marielystvej 33, 6000 Kolding, Denmark
```

```
#### Matrix duration_matrix_9_5_20 ####
     [0]  [1]  [2] 
[0]     0    4   21   |  (55.514944, 9.527195) - Marielystvej 9, 6000 Kolding, Denmark
[1]     4    0   17   |  (55.514774, 9.526439) - Marielystvej 5, 6000 Kolding, Denmark
[2]   144   12    0   |  (55.514471, 9.526256) - Marielystvej 20, 6000 Kolding, Denmark
```
Both matrices are generated with node_snap_tolerance = 0 on all waypoints.

-----------------------

Also, I have experimented with using the directions/route API instead of the matrix API. Using this API, the same two matrixes look like this (after stitching the matrix from->to elements onto a route, so to speak):

```
#### Matrix duration_matrix_106_37_33 ####
     [0]  [1]  [2] 
[0]   0.0 11.3  8.6   |  (55.515759, 9.527645) - Marielystvej 106, 6000 Kolding, Denmark
[1]  11.3  0.0  2.7   |  (55.516036, 9.525968) - Marielystvej 37, 6000 Kolding, Denmark
[2]   8.6  2.7  0.0   |  (55.515759, 9.526499) - Marielystvej 33, 6000 Kolding, Denmark
```

```
#### Matrix duration_matrix_9_5_20 ####
     [0]  [1]  [2] 
[0]   0.0  4.6 22.7   |  (55.514944, 9.527195) - Marielystvej 9, 6000 Kolding, Denmark
[1]   4.6  0.0 18.0   |  (55.514774, 9.526439) - Marielystvej 5, 6000 Kolding, Denmark
[2]  18.0 13.4  0.0   |  (55.514471, 9.526256) - Marielystvej 20, 6000 Kolding, Denmark
```
(For these matrices based on /route, node_snap_tolerance is not changed form the Valhalla default.)

So whatever is causing the very large durations in the matrix API are not present on the route API.

Also note that all coordinates used are vejpunkter (road coordinates) as previded by the Danish "Road Agency" (DAWA).

Hope this helps.

**Comment by chrstnbwnkl** [comment]
@klausfrandsen can you provide the shapes for the routes using `shape_format=polyline6`? I just came across this issue again and it seems like it could be related to [this observation](https://github.com/valhalla/valhalla/pull/4634#issuecomment-2137218436) and the test case I was able to create that recreates these weird output shapes

**Comment by klausfrandsen** [comment]
@chrstnbwnkl I do unfortunately not have those shapes easily available any longer. But I agree that it would be great to see how the routes behind the numbers actually look. Looking at #4634, it definitely looks related on the face of it.

**Comment by chrstnbwnkl** [comment]
I am going to close this issue for now, it looks like this should have been fixed in #4634, but feel free to reopen if the issue persists.
